server/src/main/java/org/opensearch/gateway/remote/RemoteClusterStateService.java

```diff
@@ -0,0 +1,371 @@
+     */
+    @Nullable
+    public ClusterMetadataMarker writeIncrementalMetadata(
+        ClusterState previousClusterState,
+        ClusterState clusterState,
+        ClusterMetadataMarker previousMarker
+    ) throws IOException {
+        final long startTimeMillis = relativeTimeMillisSupplier.getAsLong();
+        if (clusterState.nodes().isLocalNodeElectedClusterManager() == false) {
+            logger.error("Local node is not elected cluster manager. Exiting");
+            return null;
+        }
+        assert previousClusterState.metadata().coordinationMetadata().term() == clusterState.metadata().coordinationMetadata().term();
+        assert REMOTE_CLUSTER_STATE_ENABLED_SETTING.get(settings) == true : "Remote cluster state is not enabled";
+        final Map<String, Long> previousStateIndexMetadataVersionByName = new HashMap<>();
+        for (final IndexMetadata indexMetadata : previousClusterState.metadata().indices().values()) {
+            previousStateIndexMetadataVersionByName.put(indexMetadata.getIndex().getName(), indexMetadata.getVersion());
+        }
+
+        int numIndicesUpdated = 0;
+        int numIndicesUnchanged = 0;
+        final Map<String, ClusterMetadataMarker.UploadedIndexMetadata> allUploadedIndexMetadata = previousMarker.getIndices()
+            .stream()
+            .collect(Collectors.toMap(UploadedIndexMetadata::getIndexName, Function.identity()));
+        for (final IndexMetadata indexMetadata : clusterState.metadata().indices().values()) {
+            final Long previousVersion = previousStateIndexMetadataVersionByName.get(indexMetadata.getIndex().getName());
+            if (previousVersion == null || indexMetadata.getVersion() != previousVersion) {
+                logger.trace(
+                    "updating metadata for [{}], changing version from [{}] to [{}]",
+                    indexMetadata.getIndex(),
+                    previousVersion,
+                    indexMetadata.getVersion()
+                );
+                numIndicesUpdated++;
+                final String indexMetadataKey = writeIndexMetadata(
+                    clusterState.getClusterName().value(),
+                    clusterState.getMetadata().clusterUUID(),
+                    indexMetadata,
+                    indexMetadataFileName(indexMetadata)
+                );
+                final UploadedIndexMetadata uploadedIndexMetadata = new UploadedIndexMetadata(
+                    indexMetadata.getIndex().getName(),
+                    indexMetadata.getIndexUUID(),
+                    indexMetadataKey
+                );
+                allUploadedIndexMetadata.put(indexMetadata.getIndex().getName(), uploadedIndexMetadata);
+            } else {
+                numIndicesUnchanged++;
+            }
+            previousStateIndexMetadataVersionByName.remove(indexMetadata.getIndex().getName());
+        }
+
+        for (String removedIndexName : previousStateIndexMetadataVersionByName.keySet()) {
```

Comment:
nit: Removing entries from `previousStateIndexMetadataVersionByName` and using it to remove deleted indices from metadata makes things harder to understand.

What we want here is to check all the entries in previousState that are not present in currentState and remove them.

Reply:
This logic is very similar to the logic in [PersistedClusterStateService](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/gateway/PersistedClusterStateService.java#L756-L784). Wanted to keep the logic in sync.